server/blob/blob-deduplicating/src/test/scala/GCPropertiesTest.scala

```diff
@@ -0,0 +1,283 @@
+import org.apache.james.blob.api.{BlobId, TestBlobId}
+import org.scalacheck.Prop.forAll
+import org.scalacheck.{Arbitrary, Gen, Properties}
+
+case class Generation(id: Long) {
+  def previous: Generation = copy(id - 1)
+  def next: Generation = copy(id + 1)
+
+  def collectibles(targetedGeneration: Generation): Set[Generation] =
+    ((id + 1) to (targetedGeneration.id - 2)).toSet.map(Generation(_))
+}
+
+object Generation {
+  implicit val ord: Ordering[Generation] = Ordering.by(unapply)
+}
+
+case class Iteration(id: Long, processedGenerations: Set[Generation]) {
+  def next(generations: Set[Generation]): Iteration = Iteration(id + 1, generations)
+  def lastGeneration: Generation = processedGenerations.max
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(Generation(-1)))
+}
+
+case class ExternalID(id: String) // TODO
+
+sealed trait Event {
+  def blob: BlobId
+  def externalId: ExternalID
+  def generation: Generation
+}
+
+case class Reference(externalId: ExternalID, blobId: BlobId, generation: Generation) extends Event {
+  override def blob: BlobId = blobId
+}
+
+case class Deletion(generation: Generation, reference: Reference) extends Event {
+  override def blob: BlobId = reference.blob
+  override def externalId: ExternalID = reference.externalId
+}
+
+case class Report(iteration: Iteration,
+                  blobsToDelete: Set[(Generation, BlobId)]
+                 )
+
+case class State(blobs: Set[BlobId],
+                 references: Map[Generation, Set[Reference]],
+                 deletions: Map[Generation, Set[Deletion]]) {
+  private lazy val referencedBlobsAcrossGenerations: Map[Generation, ReferencedBlobs] = {
```

Comment:
I think this method is not correct because you don't take care of event ordering.
if you have: create A + delete A + create A (where A is a BlobId) you should still keep a reference to it

Reply:
This method is correct because I do not care of event ordering.

I consider all the additions first, then all the deletions, so, all deletions are done on existing blobids, which cannot be referenced afterward.

Reply:
why it can't be referenced afterward? If you didn't switch generation yet, as blobId is related to blob content, you can very well have a Create with BlobId A as I described, don't you think?

Reply:
No, you have the State at one point in times and it does not change when the GC plans.